pypeit/edgetrace.py

```diff
@@ -960,384 +976,486 @@ def initial_trace(self, bpm=None, save=False):
-        # Add PCA, if available
-        if self.pca is not None:
-            if self.par['left_right_pca']:
-                hdu += [self.pca[0].to_hdu(name='LPCA'), self.pca[1].to_hdu(name='RPCA')]
-            else:
-                hdu += [self.pca.to_hdu()]
-        # TODO: These things are going to go in slittrace.SlitTraceSet
-        if self.design is not None:
-            hdu += [fits.BinTableHDU(header=designhdr, data=self.design, name='DESIGN')]
-        if self.objects is not None: 
-            hdu += [fits.BinTableHDU(data=self.objects, name='OBJECTS')]
-        # Write the fits file; note not everything is written. Some
-        # arrays are reconstruced by the load function.
-        hdu.writeto(outfile, overwrite=True, checksum=checksum)
-
-        # Compress the file if the output filename has a '.gz'
-        # extension; this is slow but still faster than if you have
-        # astropy.io.fits do it directly
-        if compress:
-            msgs.info('Compressing file to: {0}.gz'.format(outfile))
-            io.compress_file(outfile, overwrite=overwrite)
-
-    def load(self, validate=True, rebuild_pca=False):
+    def _bundle(self):
+        """
+        Bundle the object for writing using
+        :func:`~pypeit.datamodel.DataContainer.to_hdu`.
         """
-        Load and reinitialize the trace data.
 
-        Data is read from :attr:`master_file_path` and used to
-        overwrite any internal data. Specific comparisons of the
-        saved data are performed to ensure the file is consistent
-        with having been written by a consistent version of the code;
-        see :func:`_reinit`.
+        # TODO: This will need to be edited if/when astropy.table.Table
+        # objects with the slit matching results get added to the
+        # datamodel.
 
-        To load a full :class:`EdgeTraceSet` directly from a file,
-        use :func:`from_file`.
+        # All of the datamodel elements that can be written as a header
+        # are written to the SOBELSIG extension.
+        d = []
+        for key in self.keys():
+            if self[key] is None:
+                continue
+            if isinstance(self[key], DataContainer):
+                d += [{key: self[key]}]
+                continue
+            if self.datamodel[key]['otype'] == np.ndarray:
+                # TODO: We might want a general solution in
+                # DataContainer that knows to convert (and revert)
+                # boolean arrays to np.uint8 types. Does anyone know a
+                # way around this? It's annoying that ImageHDUs cannot
+                # have boolean type arrays.
+                if issubclass(self[key].dtype.type, (bool, np.bool_)):
+                    d += [{key: self[key].astype(np.uint8)}]
+                elif issubclass(self[key].dtype.type, (float, np.floating)):
+                    d += [{key: self[key].astype(self.output_float_dtype)}]
```

Comment:
This works here because we are ok casting all the floats as `float32`.  But to handle mixed cases (e.g. `spec2d`) we may want to instead add another key to the datamodel (e.g. `ftype=np.float32`) when we define a floating type array.

Reply:
Yeah, that's a good idea and should be easy enough to implement.  Added an issue.